File: scripts/network-shares.rb
#!/usr/bin/env ruby
# Mount configured shares via CIFS or SSHFS
# Steen Hegelund
# Time-Stamp: 2022-Sep-24 19:10
#

require 'optparse'
require 'yaml'

$cmdoptions = {}
cfgfile = File.expand_path('~/.config/ruby/network-shares.yaml')
if !File.exist?(cfgfile)
    puts "Could not find #{cfgfile}"
    exit(1)
end

def get_system_mounts(shares)
  shares.each do |_name, config|
    config['systemmounts'] = {}
    next unless config['mounts']
    config['mounts'].each do |name|
      config['systemmounts'][name] = '-'
    end
  end
  resp = `mount`.split("\n")
  resp.each do |line|
    item = /\S+ on (\S+)/.match(line)
    next if item.nil?
    shares.each do |_name, config|
      next unless config['mounts']
      config['mounts'].each do |lname|
        ldir = File.expand_path(File.join(config['local'], lname))
        next unless item[1] == ldir
        config['systemmounts'][lname] = "-> #{item[1]}"
      end
    end
  end
end

def show_folders(config)
  config['mounts'].each do |name|
    case config['type']
    when 'cifs'
      rdir = "#{config['remote']}/#{name}"
    when 'sshfs'
      rdir = "#{config['options']['username']}@#{config['remote']}:#{name}"
    else
      puts "\"#{config['type']}\" mounts not supported"
      break
    end
    puts "%-20s: %-60s %s" % [name, rdir, config['systemmounts'][name]]
  end
end

def create_folders(config, mname = nil)
  config['mounts'].each do |name|
    next if mname && mname != name
    ldir = File.expand_path(File.join(config['local'], name))
    unless Dir.exist?(ldir)
      puts "creating local folder #{ldir}"
      puts "mkdir -p #{ldir}" if $cmdoptions[:verbose]
      %x(`mkdir -p #{ldir}`)
    end
  end
end

def mount_folders(config, mname = nil)
  case config['type']
  when 'cifs'
    options = ''
    config['options'].each do |key, value|
      if value
        options += "#{key}=#{value},"
      else
        options += "#{key},"
      end
    end
    options.chomp!(',')
    config['mounts'].each do |name|
      next if mname && mname != name
      ldir = File.expand_path(File.join(config['local'], name))
      rdir = "#{config['remote']}/#{name}"
      puts "mounting folder #{ldir}"
      puts "sudo mount -t cifs #{rdir} #{ldir} -o #{options}" if $cmdoptions[:verbose]
      %x(`sudo mount -t cifs #{rdir} #{ldir} -o #{options}`)
    end
  when 'sshfs'
    config['mounts'].each do |name|
      next if mname && mname != name
      rdir = "#{config['options']['username']}@#{config['remote']}:#{name}"
      ldir = File.expand_path(File.join(config['local'], name))
      puts "mounting folder #{ldir}"
      puts "sshfs #{rdir} #{ldir}" if $cmdoptions[:verbose]
      %x(`sshfs #{rdir} #{ldir}`)
    end
  else
    puts "\"#{config['type']}\" mounts not supported"
  end
end

def unmount_folders(config, mname = nil)
  config['mounts'].each do |name|
    next if mname && mname != name
    next if config['systemmounts'][name] == '-'
    ldir = File.expand_path(File.join(config['local'], name))
    puts "unmounting folder #{ldir}"
    puts "umount #{ldir}" if $cmdoptions[:verbose]
    %x(`sudo umount #{ldir}`)
  end
end

opt_parser = OptionParser.new do |opts|
  opts.banner = "Usage: #{File.basename(__FILE__)} [options]

    The default is to show the list of configured shares

Options:"
  opts.on('-v', '--verbose', 'Show actions') do |d|
    $cmdoptions[:verbose] = d
  end
  opts.on('-m', '--mount', 'Mount all listed shares') do |d|
    $cmdoptions[:mount] = d
  end
  opts.on('-u', '--unmount', 'Unmount all listed shares') do |d|
    $cmdoptions[:unmount] = d
  end
  opts.on('-o', '--mountnamed NAME', 'Mount named share') do |d|
    $cmdoptions[:mountnamed] = d
  end
  opts.on('-n', '--unmountnamed NAME', 'Unmount named share') do |d|
    $cmdoptions[:unmountnamed] = d
  end
  opts.on('-h', '--help', 'Show this message') do
    puts opts
    exit
  end
end

opt_parser.parse!(ARGV)

shares = YAML.safe_load(File.read(cfgfile))

if $cmdoptions[:mountnamed]
  shares.each do |_name, config|
    next if config['mounts'].nil?
    create_folders(config, $cmdoptions[:mountnamed])
    mount_folders(config, $cmdoptions[:mountnamed])
  end
  exit
end

if $cmdoptions[:mount]
  shares.each do |_name, config|
    next if config['mounts'].nil?
    create_folders(config)
    mount_folders(config)
  end
  exit
end

get_system_mounts(shares)

if $cmdoptions[:unmountnamed]
  shares.each do |_name, config|
    next if config['mounts'].nil?
    unmount_folders(config, $cmdoptions[:unmountnamed])
  end
  exit
end

if $cmdoptions[:unmount]
  shares.each do |_name, config|
    next if config['mounts'].nil?
    unmount_folders(config)
  end
  exit
end

shares.each do |_name, config|
  next if config['mounts'].nil?
  show_folders(config)
end

File: scripts/network-shares.py
#! /usr/bin/env python3
'''
Steen Hegelund
Time-Stamp: 2024-Oct-11 13:40

Mount configured shares via CIFS or SSHFS

The scripts reads a configuration file at ~/.config/python/network-shares.yaml
The configuration can consist of a number of sections like the example below:

    <name>:
      type: cifs
      local: ~/mnt/laptop
      remote: <remote server e.g. //server/share >
      mounts:
        <mount name>: <path on remote eg. "User/name/Downloads",  This is appended on "remote" >
      options:
        username: <remote username>
        password: <remote plaintext password>
        passcmd: <command to get the remote password e.g using gpg>
        domain: <domain name>
        uid: <local username>
        gid: <local groupname>
        iocharset: utf8
        vers: <use cifs protocol version, e.g. 1.0 or 2.0>

There can be multiple mounts specified on the same server, and you use the mount name to select which one to use.

'''
# vim: set ts=4 sw=4 sts=4 tw=120 cc=120 et ft=python :

import os.path
import sys
import argparse
import subprocess
import re
import yaml

global verbose

verbose = False


def parse_arguments():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--verbose', '-v', help='Show actions', action='store_true')
    parser.add_argument('--mountall', '-a', help='Mount all listed shares', action='store_true')
    parser.add_argument('--unmountall', '-x', help='Unmount all listed shares', action='store_true')
    parser.add_argument('--unmount', '-u', nargs='*', help='Unmount this share')
    parser.add_argument('mount', nargs='*', help='Mount this share')

    return parser.parse_args()


def syscmd(cmd):
    if verbose:
        print(f'=> {cmd}')
    cp = subprocess.run(cmd.split(), stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    if cp.returncode == 0:
        return cp.stdout.decode().split('\n')
    return []


def create_folders(config, mname=None):
    for name in config['mounts'].keys():
        if mname is not None and name not in mname:
            continue
        ldir = os.path.expanduser(os.path.join(config['local'], name))
        if os.path.exists(ldir):
            continue
        print(f'creating local folder: {ldir}')
        syscmd(f'mkdir -p {ldir}')


def get_password(cmdstr):
    cmd = []
    for elem in cmdstr.split():
        cmd.append(os.path.expanduser(elem))
    cp = subprocess.run(cmd, capture_output=True)
    lines = cp.stdout.decode().split('\n')
    for line in lines:
        if len(line) > 0:
            return line
    return ''


def update_options(config):
    for key, value in config['options'].items():
        if key == 'passcmd':
            value = get_password(value)
            if len(value):
                config['options']['password'] = value
            del config['options']['passcmd']
            break


def parse_options(config):
    options = []
    if config['type'] == 'cifs':
        update_options(config)
        for key, value in config['options'].items():
            if value:
                options.append(f"{key}={value}")
            else:
                options.append(f"{key}")
    return ','.join(options)


def mount_folders(config, mname=None):
    optstr = parse_options(config)
    for name, share in config['mounts'].items():
        if mname is not None and name not in mname:
            continue
        ldir = os.path.expanduser(os.path.join(config['local'], name))
        rdir = f"{config['remote']}/{share}"
        print(f'mounting local folder: {ldir}')
        syscmd(f'sudo mount -t cifs {rdir} {ldir} -o {optstr}')


def unmount_folders(config, mname=None):
    for name in config['mounts'].keys():
        if mname is not None and name not in mname:
            continue
        ldir = os.path.expanduser(os.path.join(config['local'], name))
        print(f'unmounting folder {ldir}')
        syscmd(f"sudo umount {ldir}")


def get_system_mounts(shares):
    for name, config in shares.items():
        config['systemmounts'] = {}
        if not config['mounts']:
            continue
        for name in config['mounts'].keys():
            config['systemmounts'][name] = '-'
    resp = syscmd('mount')
    regex = re.compile(r'\S+ on (\S+)')
    for line in resp:
        item = regex.match(line)
        if not item:
            continue
        for name, config in shares.items():
            if not config['mounts']:
                continue
            for sname, share in config['mounts'].items():
                ldir = os.path.expanduser(os.path.join(config['local'], sname))
                if item[1] != ldir:
                    continue
                config['systemmounts'][sname] = f'-> {item[1]}'


def show_folders(config):
    for name, share in config['mounts'].items():
        if config['type'] == 'cifs':
            rdir = f"{config['remote']}/{share}"
        elif config['type'] == 'sshfs':
            rdir = f"{config['options']['username']}@{config['remote']}:{share}"
        else:
            print(f'\"{config["type"]}\" mounts not supported')
            break
        print(f'{name:<20}: {rdir:<60} {config["systemmounts"][name]}')


def show_mounts(shares):
    get_system_mounts(shares)
    print('Currently mounted shares:\n')
    print(f'{"Name":<20}: {"Network Share":<60} Local Mounted Folder')
    print(f'{"-"*110}')
    for name, config in shares.items():
        if not config['mounts']:
            continue
        show_folders(config)
    sys.exit(0)


def command(args, cfgfile):
    with open(cfgfile, 'rt') as fobj:
        shares = yaml.safe_load(fobj)

        if args.mount:
            for name, config in shares.items():
                if not config['mounts']:
                    continue
                create_folders(config, args.mount)
                mount_folders(config, args.mount)
            show_mounts(shares)

        elif args.mountall:
            for name, config in shares.items():
                if not config['mounts']:
                    continue
                create_folders(config)
                mount_folders(config)
            show_mounts(shares)

        elif args.unmount:
            for name, config in shares.items():
                if not config['mounts']:
                    continue
                unmount_folders(config, args.unmount)
            show_mounts(shares)

        elif args.unmountall:
            for name, config in shares.items():
                if not config['mounts']:
                    continue
                unmount_folders(config)
            show_mounts(shares)

        else:
            show_mounts(shares)


if __name__ == '__main__':
    args = parse_arguments()
    verbose = args.verbose

    cfgfile = os.path.expanduser('~/.config/python/network-shares.yaml')
    if not os.path.exists(cfgfile):
        print(f'Could not find {cfgfile}')
        sys.exit(1)

    command(args, cfgfile)
